src/dao.cpp

```diff
@@ -97,6 +98,42 @@ namespace hypha
       proposal->close(docprop);
    }
 
+   void dao::proposepub(const name &proposer, const checksum256 &proposal_hash)
+   {
+      TRACE_FUNCTION()
+      EOS_CHECK(!isPaused(), "Contract is paused for maintenance. Please try again later.");
+
+      Document docprop(get_self(), proposal_hash);
+      name proposal_type = docprop.getContentWrapper().getOrFail(SYSTEM, TYPE)->getAs<eosio::name>();
+
+      Proposal *proposal = ProposalFactory::Factory(*this, proposal_type);
+      proposal->publish(proposer, docprop);
+   }
+
+   void dao::proposerem(const name &proposer, const checksum256 &proposal_hash)
+   {
+      TRACE_FUNCTION()
+      EOS_CHECK(!isPaused(), "Contract is paused for maintenance. Please try again later.");
+
+      Document docprop(get_self(), proposal_hash);
+      name proposal_type = docprop.getContentWrapper().getOrFail(SYSTEM, TYPE)->getAs<eosio::name>();
+
+      Proposal *proposal = ProposalFactory::Factory(*this, proposal_type);
+      proposal->remove(proposer, docprop);
+   }
+
+   void dao::proposeupd(const name &proposer, const checksum256 &proposal_hash, ContentGroups &content_groups) 
+   {
+      TRACE_FUNCTION()
+      EOS_CHECK(!isPaused(), "Contract is paused for maintenance. Please try again later.");
+
+      Document docprop(get_self(), proposal_hash);
+      name proposal_type = docprop.getContentWrapper().getOrFail(SYSTEM, TYPE)->getAs<eosio::name>();
+
+      Proposal *proposal = ProposalFactory::Factory(*this, proposal_type);
```

Comment:
other than the last line these 3 methods repeat the same code - is there a way to integrate this so this code is in one place and only one place? DRY principle (don't repeat yourself).

Reply:
It should be possible, except for the first line, which is a macro that collects the file/line for debug purposes.
I would preffer to do it on a separate PR and rebase accordingly.